flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/TempFileManager.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.filesystem;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Manage temporary files for writing files. Use special rules to organize directories
+ * for temporary files.
+ *
+ * <p>Temporary file directory contains the following directory parts:
+ *  1.temporary base path directory.
+ *  2.checkpoint id directory.
+ *  3.task id directory.
+ *  4.directories to specify partitioning.
+ *  5.data files.
+ *  eg: /tmp/cp-1/task-0/p0=1/p1=2/fileName.
+ */
+@Internal
+public class TempFileManager {
+
+	private static final String CHECKPOINT_DIR_PREFIX = "cp-";
+	private static final String TASK_DIR_PREFIX = "task-";
+
+	private final int taskNumber;
+	private final long checkpointId;
+	private final Path taskTmpDir;
+
+	private transient int nameCounter = 0;
+
+	public TempFileManager(Path temporaryPath, int taskNumber, long checkpointId) {
+		checkArgument(checkpointId != -1, "checkpoint id start with 0.");
+		this.taskNumber = taskNumber;
+		this.checkpointId = checkpointId;
+		this.taskTmpDir = new Path(
+				new Path(temporaryPath, checkpointName(checkpointId)),
+				TASK_DIR_PREFIX + taskNumber);
+	}
+
+	public Path getTaskTemporaryPath() {
+		return taskTmpDir;
+	}
+
+	/**
+	 * Generate a new path with directories.
+	 */
+	public Path generateTempFile(String... directories) throws Exception {
```

Comment:
`createPartitionDir` will be more accurate, change parameter name to `partitions`